state/src/replicationv2/snapshot.rs

```diff
@@ -0,0 +1,131 @@
+    tx.commit().map_err(ReplicationV2Error::Storage)?;
+
+    // TODO: Remove unneeded tables
+
+    // Zip the snapshot
+    let zip_task = tokio::task::spawn_blocking(move || zip_file(&snapshot_path));
+    zip_task.await.map_err(|_| ReplicationV2Error::Snapshot(ERR_AWAIT_ZIP.to_string()))?
+}
+
+/// Copy the data file of the DB
+///
+/// Returns the location at which the data was copied
+async fn make_data_copy(path: &str, out_dir: &str) -> Result<PathBuf, ReplicationV2Error> {
+    let path = PathBuf::from(path);
+    let data_path = path.join(MDBX_DATA_FILE);
+    let snapshot_path = PathBuf::from(out_dir).join(SNAPSHOT_FILE);
+
+    // Make a copy of the data
+    tokio::fs::copy(data_path, snapshot_path.clone())
+        .await
+        .map_err(err_str!(ReplicationV2Error::Snapshot))?;
+    Ok(snapshot_path)
+}
+
+/// Zip a file and delete the original
+fn zip_file(path: &PathBuf) -> Result<(), ReplicationV2Error> {
+    let source_file = File::open(path).map_err(err_str!(ReplicationV2Error::Snapshot))?;
+    let mut source_reader = BufReader::new(source_file);
+    let zip_path = path.with_extension("gz");
+    let zip_file = File::create(zip_path).map_err(err_str!(ReplicationV2Error::Snapshot))?;
+
+    // gzip the file
+    let mut encoder = GzEncoder::new(zip_file, Compression::best());
+    std::io::copy(&mut source_reader, &mut encoder)
+        .map_err(err_str!(ReplicationV2Error::Snapshot))?;
+    encoder.finish().map_err(err_str!(ReplicationV2Error::Snapshot))?;
+
+    // Delete the file
+    fs::remove_file(path).map_err(err_str!(ReplicationV2Error::Snapshot))?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::env::temp_dir;
+
+    use flate2::bufread::GzDecoder;
+
+    use crate::{storage::db::DbConfig, test_helpers::tmp_db_path};
+
+    use super::*;
+
+    /// Recover a data file from a snapshot archive
+    ///
+    /// Returns the location of the recovered data file
+    async fn recover_data_from_snapshot(snapshot_path: &str) -> String {
+        let snapshot_path = PathBuf::from(snapshot_path);
+        let snapshot_path = snapshot_path.join(SNAPSHOT_FILE).with_extension("gz");
+
+        let tmp_dir = temp_dir();
+        let out_path = tmp_dir.join(MDBX_DATA_FILE);
+        let mut out_file = File::create(out_path.clone()).unwrap();
+
+        // Load the snapshot and unzip it
+        let snapshot_file = File::open(snapshot_path).unwrap();
+        let snapshot_reader = BufReader::new(snapshot_file);
+        let mut snapshot_decoder = GzDecoder::new(snapshot_reader);
+        std::io::copy(&mut snapshot_decoder, &mut out_file).unwrap();
+
+        out_path.to_str().unwrap().to_string()
+    }
+
+    /// Tests snapshotting and recovering a DB from a snapshot
+    #[tokio::test]
+    async fn test_snapshot() {
+        const TABLE: &str = "test-table";
+        let (k, v) = ("key".to_string(), "value".to_string());
+
+        let db_path = tmp_db_path();
+        let snap_path = db_path.clone();
```

Comment:
it looks like we save the snapshot to the same location as the db path, this feels like unintended usage - perhaps we make `snap_path` a different temp file?

Reply:
for tests its fine, wanted to avoid having to create new directories. it should not matter whether it is the same directory